Feature/aten nys 71  automated comment closing session year

nathanielwoodland opened the pull request:
Dev note: implemented cron hook in nys_comment module instead of nys_bills module with an eye towards similar comment automation in other content types. Open to feedback if this doesn't match business rules.

routinet commented on web/modules/custom/nys_comment/nys_comment.module:
```diff
@@ -175,3 +175,36 @@ function nys_comment_deploy_default_rejected_value() {
+  $query->fields('ns', ['field_ol_session_value']);
+  $query->condition('nc.field_comments_status', '1', '!=');
+  $query->condition('n.type', $content_types, 'IN');
+  $query->condition('ns.field_ol_session_value', $current_session_year, '<=');
```
I think this needs to be `<` rather than `<=`.  If the current session is 2023, any bills from 2023/4 should still be allowed to receive comments.

routinet reviewed (commented):
I'm wondering about two alternate approaches: 

1. Using `loadByProperties()` instead of a direct query to find targeted nodes.
2. When closing comments, using a single query for all detected nodes instead of one query for each.  Additionally, sending all affected `$nid` values through `resetCache()` with a single call.

routinet commented on web/modules/custom/nys_comment/nys_comment.module:
```diff
@@ -175,3 +175,36 @@ function nys_comment_deploy_default_rejected_value() {
+  $query->join('node__field_comments', 'nc', 'n.nid = nc.entity_id');
+  $query->fields('n', ['nid', 'type']);
+  $query->fields('ns', ['field_ol_session_value']);
+  $query->condition('nc.field_comments_status', '1', '!=');
```
Search should limit itself to `nc.field_comments_status = 2`.  If someone has marked comments as hidden, it should stay that way until manually changed.

routinet commented:
Please add a comment to document: 

- Bills and resolutions are the only content types allowing comments at this time.
- The dependency on `field_ol_session` implicitly limits this to bill/resolution nodes anyways.

routinet commented:
I really ought to collect all of my thoughts before spamming comments on a GH thread.  AND another thing...

The hook is depending on the NY Senate config point for current session.  Realistically, this can work just as well through simple date detection.  In either case, I do not think a normal cron run is necessary here; conditions to trigger an automated closings for comments will change literally once every two years.  If we continue to use the config point, perhaps execute this as a result of changing it?

A bit off topic, but same goes for the config point.  Is it really necessary?  Is there a reason to have that, vs. a simple date detection method in a module/service?

cc @fksaintil @aheaphy

nathanielwoodland commented:
Okay @routinet, I'm ready for you to carve this up once again! 😃 

Here are the updates relevant to your PR review:

1. Use "<" instead of "<=" operator (yikes, good catch!)
2. Operate on both "bill" and "resolution" records
3. Limit results to records where field_comments_status = 2
4. Close comments / reset caches in batch operations, not in loop (forgive for being a bit loopy from lack of sleep ...)
5. Updated cron config to run only once per month (this was always my plan once testing was completed; we can make this even less frequent)

Additional updates from my testing:

1. I discovered that many old (maybe migrated) resolution records didn't have their field_comment data properly instantiated, so I added code to find those records and insert the pre-req data (I did this in the cron hook, vs. a one-time install hook, as I'm not certain if there's code that may create new resolution records with missing comment DB data).

Things to note:

1. I kept this in DB queries instead of entity queries (e.g. loadByProperties) to avoid timeout issues, as we're dealing with almost 200,000 records, and loading and saving all those entities in a loop can take significant time (especially if there are node update hooks that get called). The main risk comes in the comment field name changing, but that would also break an entity query.
2. Happy to discuss the session config point. Leaving as-is, for now.

routinet commented:
After some investigation, I have confirmed that new bills/resolutions imported from OpenLeg receive the expected row, so we should not need to execute the correction on a regular basis.  I'm assuming the lack of related comment field rows is due to the bill import being completed prior to the installation of the comment module, or some other coincidence of timing in its implementation.  Should this problem reoccur, we can address it at that time.  

Refactor the correction to run once only.  An update hook should work well.  The entire correction algorithm can be rewritten as: ```$subquery=\Drupal::database()->select('node_field_data', 'n')
  // Probably replace with $content_types
  ->condition('n.type', ['bill', 'resolution'], 'IN')
  ->condition('nc.entity_id', '', 'IS NULL');
$subquery->leftJoin('node__field_comments', 'nc', 'nc.entity_id=n.nid AND n.type=nc.bundle');
$subquery->addField('n', 'type', 'bundle');
$subquery->addExpression("'0'", 'deleted');
$subquery->addField('n', 'nid', 'entity_id');
$subquery->addField('n', 'vid', 'revision_id');
$subquery->addExpression("'en'", 'langcode');
$subquery->addExpression("'0'", 'delta');
$subquery->addExpression("'2'", 'field_comments_status');
\Drupal::database()->insert('node__field_comments')
  ->from($subquery)
  ->execute();```

nathanielwoodland commented:
@routinet, I've addressed your latest feedback and re-tested locally. Feel free to check this again.
